plugins/actsvg_visualization/include/detray/plugins/actsvg_visualization/proto/utils/surface_functors.hpp

```diff
@@ -0,0 +1,371 @@
+            if (surface.is_portal()) {
+                if (auto radius =
+                        surface.template visit_mask<outer_radius_functor>()) {
+                    if (*radius > mask[mask.get_shape().e_r]) {
+                        return surface_normal;
+                    }
+                }
+            }
+        }
+        return typename detector_t::scalar_type{-1} * surface_normal;
+    }
+};
+
+/// @brief A functor to set the proto surfaces type and bounds to be equivalent
+/// to the mask.
+struct to_proto_surface_functor {
+
+    public:
+    template <typename mask_group_t, typename index_t, typename detector_t>
+    DETRAY_HOST inline auto operator()(
+        const mask_group_t& mask_group, const index_t& index,
+        const typename detector_t::geometry_context& context,
+        const detray::surface<detector_t>& d_surface) const {
+        const auto& m = mask_group[index];
+        return to_proto_surface(context, d_surface, m.get_shape(), m.values());
+    }
+
+    private:
+    /// @brief Returns the proto surface for remaining shapes.
+    template <typename detector_t, typename bounds_t, typename polygon_t>
+    auto inline to_proto_surface(
+        const typename detector_t::geometry_context& context,
+        const detray::surface<detector_t>& d_surface, const polygon_t,
+        const bounds_t&) const {
+        proto::proto_surface p_surface;
+        p_surface._type = proto::proto_surface::type::e_polygon;
+        std::array dir{0., 0., 1.};
+        set_vertices(p_surface, d_surface.global_vertices(context, dir));
+        return p_surface;
+    }
+
+    /// @brief Returns the proto surface for 2D cylinders.
+    template <typename detector_t, typename bounds_t, bool kRadialCheck,
+              template <typename> class intersector_t>
+    auto inline to_proto_surface(
+        const typename detector_t::geometry_context& context,
+        const detray::surface<detector_t>& d_surface,
+        const detray::cylinder2D<kRadialCheck, intersector_t>& shape,
+        const bounds_t& bounds) const {
+        // Rotation for circular objects is currently not supported.
+        assert(d_surface.transform.rotation(context) ==
+               typename detector_t::transform3{});
+        // Only translation one z axis is supported.
+        assert(d_surface.transform.translate(context).x() == 0 &&
+               d_surface.transform.translate(context).y() == 0);
+
+        proto::proto_surface p_surface;
+        auto r = static_cast<actsvg::scalar>(bounds[shape.e_r]);
+        auto nhz = static_cast<actsvg::scalar>(bounds[shape.e_n_half_z]);
+        auto phz = static_cast<actsvg::scalar>(bounds[shape.e_p_half_z]);
+        p_surface._type = proto::proto_surface::type::e_cylinder;
+        p_surface._radii = {static_cast<actsvg::scalar>(0), r};
+        auto hz = (phz - nhz) / 2 +
+                  static_cast<actsvg::scalar>(d_surface.center(context)[2]);
+        p_surface._zparameters = {nhz + hz, hz};
```

Comment:
What exactly are the ```_zparameters``` ? Is it not just 
```suggestion
        p_surface._zparameters = {nhz, phz};
```